tests/unit/test_relational_config.py

```diff
@@ -0,0 +1,101 @@
+    assert all(all(condition for condition in pair) for pair in pairs)
+
+
+def test_no_relations(otelcol_charm):
+    """Scenario: Direct signals to debug if no data sink exists."""
+    # GIVEN no relations
+    ctx = Context(otelcol_charm)
+    state_in = State(containers=[Container(name="otelcol", can_connect=True)])
+    # WHEN any event executes the reconciler
+    state_out = ctx.run(ctx.on.update_status(), state_in)
+    otelcol = state_out.get_container("otelcol")
+    # THEN the otelcol service has started
+    assert otelcol.services["otelcol"].is_running()
+    fs = otelcol.get_filesystem(ctx)
+    otelcol_config = fs.joinpath(*otelcol_charm._config_path.strip("/").split("/"))
+    # AND the otelcol config was pushed to the workload container
+    assert otelcol_config.exists()
+    cfg = yaml.safe_load(otelcol_config.read_text())
+    # AND the pipelines are valid
+    check_valid_pipelines(cfg)
+
+
+def test_loki_exporter(otelcol_charm):
+    """Scenario: Fan-out logging architecture."""
+    # GIVEN a relation to multiple Loki units
+    remote_units_data = {
+        0: {"endpoint": '{"url": "http://fqdn-0:3100/loki/api/v1/push"}'},
+        1: {"endpoint": '{"url": "http://fqdn-1:3100/loki/api/v1/push"}'},
+    }
+    data_sink = Relation(
+        endpoint="send-loki-logs", interface="loki_push_api", remote_units_data=remote_units_data
+    )
+    ctx = Context(otelcol_charm)
+    container = Container(name="otelcol", can_connect=True)
+    state_in = State(
+        relations=[data_sink],
+        containers=[container],
+    )
+    # WHEN any event executes the reconciler
+    state_out = ctx.run(ctx.on.update_status(), state_in)
+    otelcol = state_out.get_container("otelcol")
+    # THEN the otelcol service has started
+    assert otelcol.services["otelcol"].is_running()
+    fs = otelcol.get_filesystem(ctx)
+    otelcol_config = fs.joinpath(*otelcol_charm._config_path.strip("/").split("/"))
+    # AND the otelcol config was pushed to the workload container
+    assert otelcol_config.exists()
+    cfg = yaml.safe_load(otelcol_config.read_text())
+    # AND one exporter per loki unit in relation exists in the config
+    prom_exporters = [f"loki/{idx}" for idx in range(len(remote_units_data))]
+    assert set(prom_exporters).issubset(set(cfg["exporters"].keys()))
+    # AND the pipelines are valid
+    check_valid_pipelines(cfg)
+
+
+def test_prometheus_exporter(otelcol_charm):
+    """Scenario: Fan-out remote writing architecture."""
+    # GIVEN a relation to multiple Prometheus units
+    remote_units_data = {
+        0: {"remote_write": '{"url": "http://fqdn-0:9090/api/v1/write"}'},
+        1: {"remote_write": '{"url": "http://fqdn-1:9090/api/v1/write"}'},
+    }
+    data_sink = Relation(
+        endpoint="send-remote-write",
+        interface="prometheus_remote_write",
+        remote_units_data=remote_units_data,
+    )
+    ctx = Context(otelcol_charm)
```

Comment:
Sorry for creating rebase conflict for you, but:
- `ctx` is now a fixture.
- `otelcol_charm` is gone and instead of `otelcol_charm._config_path` you can use `constants.CONFIG_PATH`.
- `execs` is a new fixture, and you need to pass it as arg to `Container(...)`.